### PR TITLE
ui: added node constraint solver to UI

### PR DESCRIPTION
**Pull request by d4l3k**
This builds on the rest of the allocator PRs (#8786). Only review the last commit in this PR.

cc @maxlang @BramGruneir

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/8887)

### CODE REVIEW

**Comment by maxlang** [comment]
:lgtm:

---

Review status: 0 of 54 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[ui/app/containers/nodeConstraints.tsx, line 20 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQMiDIJUBCgoKa4_BIx:-KQMiDIK2A0DdLvmZJbi:b-euhjx9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d482d0359d77e16bb61d3c25fe46e96d4140bb79/ui/app/containers/nodeConstraints.tsx#L20)):*
> ```XML
> 
> interface ConstraintsData {
>   state: KeyedCachedDataReducerState<cockroach.server.serverpb.ConstraintsResponseMessage>;
> ```

State is a pretty overloaded word, since we already have the Redux state and the React component state.

---

*[ui/app/containers/nodeConstraints.tsx, line 32 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQMibzcY_xqwob1lcvc:-KQMibzcY_xqwob1lcvd:b-c27g3m) ([raw file](https://github.com/cockroachdb/cockroach/blob/d482d0359d77e16bb61d3c25fe46e96d4140bb79/ui/app/containers/nodeConstraints.tsx#L32)):*
> ```XML
> 
> interface ConstraintsState {
>   constraints: string;
> ```

I feel like this could be hard to follow, since this isn't actually a list of constraints but more of a constraint selector of some sort?

---

*[ui/app/containers/nodeConstraints.tsx, line 39 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQMjMwUOTCZeq79QKog:-KQMjMwVEFLck9XxzCUG:b-mdxl5u) ([raw file](https://github.com/cockroachdb/cockroach/blob/d482d0359d77e16bb61d3c25fe46e96d4140bb79/ui/app/containers/nodeConstraints.tsx#L39)):*
> ```XML
>  * Constraints component.
>  */
> type ConstraintsProps = ConstraintsData & ConstraintsActions;
> ```

Since the types here a pretty simple, you can just combine everything straight up, without composing it piecewise.

---

*[ui/app/containers/nodeConstraints.tsx, line 59 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQMh5qi2jZN5XU3riAm:-KQMh5qi2jZN5XU3riAn:b-e3r9m7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d482d0359d77e16bb61d3c25fe46e96d4140bb79/ui/app/containers/nodeConstraints.tsx#L59)):*
> ```XML
>   componentDidMount() {
>     this.refreshConstraints();
>     const node = ReactDom.findDOMNode(this);
> ```

@mrtracy Do you prefer using `ReactDom.findDOMNode(this);` or the ref technique you used in `linegraph.tsx`?

---

*[ui/app/containers/nodeConstraints.tsx, line 61 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQMhlsSjFvfZ7Oy2Bo1:-KQMhlsSjFvfZ7Oy2Bo2:b-7d6bua) ([raw file](https://github.com/cockroachdb/cockroach/blob/d482d0359d77e16bb61d3c25fe46e96d4140bb79/ui/app/containers/nodeConstraints.tsx#L61)):*
> ```XML
>     const node = ReactDom.findDOMNode(this);
>     this.chart = d3.select(node).select("svg");
>     setTimeout(this.updateChart.bind(this), 1);
> ```

Why do you need to wait for the next tick here? Also you probably don't need to specify an actual delay. You can leave off that parameter.

---

*[ui/app/containers/nodeConstraints.tsx, line 77 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQMoBahFXzYhgE7zofR:-KQMoBahFXzYhgE7zofS:b-arilnf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d482d0359d77e16bb61d3c25fe46e96d4140bb79/ui/app/containers/nodeConstraints.tsx#L77)):*
> ```XML
>     for (let candidate of state.data.candidates) {
>       candidates[candidate.desc.store_id] = candidate;
>     }
> ```

Some of this can be rewritten nicely with `lodash`.

`const candidates = _.keyBy(state.data.candidates, "desc.store_id")` 

or, to maintain type safety, 

`const candidates = _.keyBy(state.data.candidates, (c)=>c.desc.store_id)` 

---

*[ui/app/containers/nodeConstraints.tsx, line 79 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQMrNgi2bwzaUoIkfwj:-KQMrNgj_nya9DxU2ACE:b-s75z4j) ([raw file](https://github.com/cockroachdb/cockroach/blob/d482d0359d77e16bb61d3c25fe46e96d4140bb79/ui/app/containers/nodeConstraints.tsx#L79)):*
> ```XML
>     }
> 
>     const nodes: {[id: number]: cockroach.roachpb.StoreDescriptor[]} = {};
> ```

`nodes = _.groupBy(state.state.stores, (s) => s.node.node_id)`

---

*[ui/app/containers/nodeConstraints.tsx, line 91 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQMziSMAbxqtjlj4YqB:-KQMziSMAbxqtjlj4YqC:bjm6sud) ([raw file](https://github.com/cockroachdb/cockroach/blob/d482d0359d77e16bb61d3c25fe46e96d4140bb79/ui/app/containers/nodeConstraints.tsx#L91)):*
> ```XML
>     const datums: {[name: string]: Datum} = {};
>     const root: Datum = {name: "Cluster", children: []};
>     _.each(nodes, (stores: cockroach.roachpb.StoreDescriptor[], id: string) => {
> ```

I feel like this big `_.each` should be split into two parts. First map across each node, creating a node with name, attrs, children and tiers, then insert each one into the tree, removing tiers if necessary.

---

*[ui/app/containers/nodeConstraints.tsx, line 92 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQMyF7CAg4rjNmLYmoY:-KQMyF7D0T11VzuwDNwk:b-mwxtsn) ([raw file](https://github.com/cockroachdb/cockroach/blob/d482d0359d77e16bb61d3c25fe46e96d4140bb79/ui/app/containers/nodeConstraints.tsx#L92)):*
> ```XML
>     const root: Datum = {name: "Cluster", children: []};
>     _.each(nodes, (stores: cockroach.roachpb.StoreDescriptor[], id: string) => {
>       const nodeDesc = stores[0].node;
> ```

Why are you only ever using `stores[0]`?

---

*[ui/app/containers/nodeConstraints.tsx, line 121 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQMv335MnTD6P0h3B_z:-KQMv335MnTD6P0h3Ba-:btxn60i) ([raw file](https://github.com/cockroachdb/cockroach/blob/d482d0359d77e16bb61d3c25fe46e96d4140bb79/ui/app/containers/nodeConstraints.tsx#L121)):*
> ```XML
>         }
>         prefix = tierName;
>         container = c;
> ```

What about writing this recursively so you aren't setting external values in a `foreach` callback?

---

*[ui/app/containers/nodeConstraints.tsx, line 136 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQMm0KuchwYE5rb4E93:-KQMm0KvzIrzF7QMtoOs:b-yfcnq5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d482d0359d77e16bb61d3c25fe46e96d4140bb79/ui/app/containers/nodeConstraints.tsx#L136)):*
> ```XML
> 
>   updateChart() {
>     const root = this.structuredStores();
> ```

I'm not sure if you borrowed any of this code from a d3 example, but you should attribute it if you did.

---

*[ui/app/containers/nodeConstraints.tsx, line 174 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQN457fzKEMEAzVwtVl:-KQMn9cC5hPSY30oj2mk:bt833a4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d482d0359d77e16bb61d3c25fe46e96d4140bb79/ui/app/containers/nodeConstraints.tsx#L174)):*
> ```XML
>     gnode.append("circle")
>       .attr("r", 4.5)
>       .style("fill", (d: Datum) => d.candidate ? "green" : "black");
> ```

Eventually we should update this chart to conform to the styles of the rest of the site.

---

*[ui/app/util/api.ts, line 200 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQMeaQR2BjceJJNO0q3:-KQMeaQR2BjceJJNO0q4:bbwn6i) ([raw file](https://github.com/cockroachdb/cockroach/blob/d482d0359d77e16bb61d3c25fe46e96d4140bb79/ui/app/util/api.ts#L200)):*
> ```XML
> 
> export function getConstraints(req: ConstraintsRequestMessage): Promise<ConstraintsResponseMessage> {
>   // NB: raftDebug intentionally does not pass a timeout through.
> ```

Change comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-:-KQN4TNBBJAa1XRTLMT0:bnfp4nl)*

**Comment by maxlang** [comment]
cc @mrtracy

**Comment by maxlang** [comment]
Also: looks awesome!

**Comment by d4l3k** [comment]
Thanks!

---

Review status: 0 of 54 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[ui/app/containers/nodeConstraints.tsx, line 20 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQMiDIJUBCgoKa4_BIx:-KQN_D7mWSU2HEBYpCW2:b4djgbe) ([raw file](https://github.com/cockroachdb/cockroach/blob/d482d0359d77e16bb61d3c25fe46e96d4140bb79/ui/app/containers/nodeConstraints.tsx#L20)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> State is a pretty overloaded word, since we already have the Redux state and the React component state.
</details>
True, but this is literally the KeyedCachedDataReducer*State*.

---

*[ui/app/containers/nodeConstraints.tsx, line 32 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQMibzcY_xqwob1lcvc:-KQN_Q1qSIF0MGCZ3sH2:b-pab0dr) ([raw file](https://github.com/cockroachdb/cockroach/blob/d482d0359d77e16bb61d3c25fe46e96d4140bb79/ui/app/containers/nodeConstraints.tsx#L32)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> I feel like this could be hard to follow, since this isn't actually a list of constraints but more of a constraint selector of some sort?
</details>
It is the constraints, just the shorthand notation for them. It's the same format that is used when specifying ZoneConfig on the command line.

https://github.com/cockroachdb/cockroach/blob/develop/docs/RFCS/expressive_zone_config.md#examples 

---

*[ui/app/containers/nodeConstraints.tsx, line 39 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQMjMwUOTCZeq79QKog:-KQNa0BHCqz6h429pjZ-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d482d0359d77e16bb61d3c25fe46e96d4140bb79/ui/app/containers/nodeConstraints.tsx#L39)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> Since the types here a pretty simple, you can just combine everything straight up, without composing it piecewise.
</details>
Done.

---

*[ui/app/containers/nodeConstraints.tsx, line 61 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQMhlsSjFvfZ7Oy2Bo1:-KQNaA-Xzk_nbDSYbmQM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d482d0359d77e16bb61d3c25fe46e96d4140bb79/ui/app/containers/nodeConstraints.tsx#L61)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> Why do you need to wait for the next tick here? Also you probably don't need to specify an actual delay. You can leave off that parameter.
</details>
Done.

---

*[ui/app/containers/nodeConstraints.tsx, line 77 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQMoBahFXzYhgE7zofR:-KQNdutjJdmfzfgVZzhl:b85exrh) ([raw file](https://github.com/cockroachdb/cockroach/blob/d482d0359d77e16bb61d3c25fe46e96d4140bb79/ui/app/containers/nodeConstraints.tsx#L77)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> Some of this can be rewritten nicely with `lodash`.
> 
> `const candidates = _.keyBy(state.data.candidates, "desc.store_id")` 
> 
> or, to maintain type safety, 
> 
> `const candidates = _.keyBy(state.data.candidates, (c)=>c.desc.store_id)` 
</details>
Lodash is amazing. Done.

---

*[ui/app/containers/nodeConstraints.tsx, line 79 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQMrNgi2bwzaUoIkfwj:-KQNdtnZPz7P4fnZmCBQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d482d0359d77e16bb61d3c25fe46e96d4140bb79/ui/app/containers/nodeConstraints.tsx#L79)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> `nodes = _.groupBy(state.state.stores, (s) => s.node.node_id)`
</details>
Done.

---

*[ui/app/containers/nodeConstraints.tsx, line 91 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQMziSMAbxqtjlj4YqB:-KQNyglPH5TLBDk3MpY9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d482d0359d77e16bb61d3c25fe46e96d4140bb79/ui/app/containers/nodeConstraints.tsx#L91)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> I feel like this big `_.each` should be split into two parts. First map across each node, creating a node with name, attrs, children and tiers, then insert each one into the tree, removing tiers if necessary.
</details>
Done.

---

*[ui/app/containers/nodeConstraints.tsx, line 92 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQMyF7CAg4rjNmLYmoY:-KQNaMSsfsEaW8qDcJKz:b-vfjt7o) ([raw file](https://github.com/cockroachdb/cockroach/blob/d482d0359d77e16bb61d3c25fe46e96d4140bb79/ui/app/containers/nodeConstraints.tsx#L92)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> Why are you only ever using `stores[0]`?
</details>
We use stores in `_.map(stores,` a little bit further down. We're using store[0] here because all stores on a node will have the same node descriptor and we only need to access one of the redundant copies.

---

*[ui/app/containers/nodeConstraints.tsx, line 121 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQMv335MnTD6P0h3B_z:-KQNaiLtuS5iBUD5PKwg:b5i5se7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d482d0359d77e16bb61d3c25fe46e96d4140bb79/ui/app/containers/nodeConstraints.tsx#L121)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> What about writing this recursively so you aren't setting external values in a `foreach` callback?
</details>
Done with a recursive build and merge.

---

*[ui/app/containers/nodeConstraints.tsx, line 136 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQMm0KuchwYE5rb4E93:-KQNbHNLhDposdiULa3K:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d482d0359d77e16bb61d3c25fe46e96d4140bb79/ui/app/containers/nodeConstraints.tsx#L136)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> I'm not sure if you borrowed any of this code from a d3 example, but you should attribute it if you did.
</details>
Done.

---

*[ui/app/util/api.ts, line 200 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQMeaQR2BjceJJNO0q3:-KQNbT1ybLsDusK1gX5l:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d482d0359d77e16bb61d3c25fe46e96d4140bb79/ui/app/util/api.ts#L200)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> Change comment.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-:-KQN_9fapyxH3tVTYm6a:b-ucfknh)*

**Comment by mrtracy** [comment]
Review status: 0 of 54 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[ui/app/containers/nodeConstraints.tsx, line 59 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQMh5qi2jZN5XU3riAm:-KQRlRvVZi81iW1RjXbM:bm7kvwt) ([raw file](https://github.com/cockroachdb/cockroach/blob/d482d0359d77e16bb61d3c25fe46e96d4140bb79/ui/app/containers/nodeConstraints.tsx#L59)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> @mrtracy Do you prefer using `ReactDom.findDOMNode(this);` or the ref technique you used in `linegraph.tsx`?
</details>
ref is preferred whenever possible https://facebook.github.io/react/docs/more-about-refs.html#a-complete-example

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887)*

**Comment by d4l3k** [comment]
Review status: 0 of 54 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[ui/app/containers/nodeConstraints.tsx, line 59 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQMh5qi2jZN5XU3riAm:-KQS2iMaP3-lfxNxR5-P:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d482d0359d77e16bb61d3c25fe46e96d4140bb79/ui/app/containers/nodeConstraints.tsx#L59)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> ref is preferred whenever possible https://facebook.github.io/react/docs/more-about-refs.html#a-complete-example
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887)*

**Comment by maxlang** [comment]
:lgtm_strong: 

---

Review status: 0 of 54 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[ui/app/containers/nodeConstraints.tsx, line 77 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQMoBahFXzYhgE7zofR:-KQSLLQq2xKb_HCBxL7F:b1w6ws3) ([raw file](https://github.com/cockroachdb/cockroach/blob/d482d0359d77e16bb61d3c25fe46e96d4140bb79/ui/app/containers/nodeConstraints.tsx#L77)):*
<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> Lodash is amazing. Done.
</details>
Haha, yes it is.

---

*[ui/app/containers/nodeConstraints.tsx, line 121 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQMv335MnTD6P0h3B_z:-KQWa6ze18Hzw8pxrHkV:b2x6m1n) ([raw file](https://github.com/cockroachdb/cockroach/blob/d482d0359d77e16bb61d3c25fe46e96d4140bb79/ui/app/containers/nodeConstraints.tsx#L121)):*
<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> Done with a recursive build and merge.
</details>
I like this a lot better.

---

*[ui/app/containers/nodeConstraints.tsx, line 90 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQWYGmt6hggsUADKlVr:-KQWYGmt6hggsUADKlVs:b-kmp5t7) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea7afd3f1f842740ea7068fe574045e339ffae3c/ui/app/containers/nodeConstraints.tsx#L90)):*
> ```XML
>   mergeDatum(base: Datum, addition: Datum) {
>     for (let d of base.children) {
>       if (d.name === addition.name) {
> ```

This feels like a `_.find`?

`const d = _.find(base.children, {name: addition.name})` or `const d = _.find(base.children, (c) => c.name === addition.name)`
Then
```
if (d) { 
  _.each(addition.children, (c)=> this.mergeDatum(d, c)) 
} else {
   base.children.push(addition);
}

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-:-KQWaO401EiyxRHeDsjL:b-uwkowd)*
<!-- Sent from Reviewable.io -->

**Comment by d4l3k** [comment]
Review status: 0 of 54 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[ui/app/containers/nodeConstraints.tsx, line 90 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887#-KQWYGmt6hggsUADKlVr:-KQWricA25AAfwu41YG1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea7afd3f1f842740ea7068fe574045e339ffae3c/ui/app/containers/nodeConstraints.tsx#L90)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> This feels like a `_.find`?
> 
> `const d = _.find(base.children, {name: addition.name})` or `const d = _.find(base.children, (c) => c.name === addition.name)`
> Then
> ```
> if (d) { 
>   _.each(addition.children, (c)=> this.mergeDatum(d, c)) 
> } else {
>    base.children.push(addition);
> }
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8887)*

**Comment by d4l3k** [comment]
Fixed an out of bounds error and added showing errors to the UI.

**Comment by vivekmenezes** [comment]
reopen if you still care about this PR

**Comment by tamird** [comment]
I think we want this. @spencerkimball?

I doubt @d4l3k will finish it up, though.

**Comment by spencerkimball** [comment]
Cool UI, but I don't believe we need it right now.

**Comment by BramGruneir** [comment]
@piyush-singh and @vilterp for inspiration.

**Comment by vilterp** [comment]
Thanks for calling this out @BramGruneir. I see something like this fitting into an end-to-end zone config editing experience integrated into the data distribution page (some discussion of this in #6306) as well:

1. see how your data is distributed now
2. get a preview of how it would be distributed if you changed config in a certain way (what this PR seems to be doing)
3. confirm that you want to make a change
4. see rebalancing progress

[This prototype](https://github.com/vilterp/replica-matrix) ([video](https://www.youtube.com/watch?v=e0DzULki3IQ)) was my idea of how such a UI would work.
